backend/src/graphql/resolver/AdminResolver.ts

```diff
@@ -478,99 +478,116 @@ export class AdminResolver {
     @Arg('id', () => Int) id: number,
     @Ctx() context: Context,
   ): Promise<boolean> {
-    const contribution = await DbContribution.findOne(id)
-    if (!contribution) {
-      logger.error(`Contribution not found for given id: ${id}`)
-      throw new Error('Contribution not found to given id.')
-    }
-    const moderatorUser = getUser(context)
-    if (moderatorUser.id === contribution.userId) {
-      logger.error('Moderator can not confirm own contribution')
-      throw new Error('Moderator can not confirm own contribution')
-    }
-    const user = await dbUser.findOneOrFail(
-      { id: contribution.userId },
-      { withDeleted: true, relations: ['emailContact'] },
-    )
-    if (user.deletedAt) {
-      logger.error('This user was deleted. Cannot confirm a contribution.')
-      throw new Error('This user was deleted. Cannot confirm a contribution.')
-    }
-    const creations = await getUserCreation(contribution.userId, false)
-    validateContribution(creations, contribution.amount, contribution.contributionDate)
-
-    const receivedCallDate = new Date()
-
-    const queryRunner = getConnection().createQueryRunner()
-    await queryRunner.connect()
-    await queryRunner.startTransaction('REPEATABLE READ') // 'READ COMMITTED')
-    try {
-      const lastTransaction = await queryRunner.manager
-        .createQueryBuilder()
-        .select('transaction')
-        .from(DbTransaction, 'transaction')
-        .where('transaction.userId = :id', { id: contribution.userId })
-        .orderBy('transaction.balanceDate', 'DESC')
-        .getOne()
-      logger.info('lastTransaction ID', lastTransaction ? lastTransaction.id : 'undefined')
-
-      let newBalance = new Decimal(0)
-      let decay: Decay | null = null
-      if (lastTransaction) {
-        decay = calculateDecay(
-          lastTransaction.balance,
-          lastTransaction.balanceDate,
-          receivedCallDate,
-        )
-        newBalance = decay.balance
+    let retryCount = 0
```

Comment:
I am against this implementation - its the task of the database to handle locks and retries - the database handles this when properly locked and resolves locks internally.

TypeORM offers the ability to lock a row or table when using the `QueryRunner`
See: https://stackoverflow.com/questions/69012855/typeorm-database-lock-please-explain-how-to-use-database-lock-in-typeorm-using

Reply:
`SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction`

The above simply means the transaction has reached the innodb_lock_wait_timeout while waiting to obtain an exclusive lock which defaults to 50 seconds. 

Source: https://severalnines.com/blog/how-fix-lock-wait-timeout-exceeded-error-mysql/

This is what suppose to happen if we lock the tables right - and 50sec is plenty in my opinion. As said, the database handles this internally - to my knowledge - means it retries for 50sec to get a lock.

Reply:
On my opinion your solution will only technically solve the deadlock by a simple retry of the blocked transaction on database level. 

But in our case the retry on database will write incorrect data, because the business logic of "confirmContribution" needs to know the last written and committed transaction including the current balance.

The test case starts two confirmContribution invocations, which both reads the same last gradidoTransaction with its balance. The first one wins and writes correct values. If the database will repeate internally the blocked transaction without rereading the previous commited data, the second transaction will write corrupt data. The second confirmContribution will create a gradidoTransaction with the same "last transaction id" than the previous commited gradidoTransaction.

If you want to avoid such situations proactive, you need other mechanism like optimistic lock or blocking monitors to synchronize concurrent processing of the same business logic.

Reply:
In your link i found the following description:

The [InnoDB documentation](https://dev.mysql.com/doc/refman/8.0/en/innodb-parameters.html#sysvar_innodb_rollback_on_timeout) clearly says “InnoDB rolls back only the last statement on a transaction timeout by default”. In this case, we do not get the transaction atomicity offered by InnoDB. The atomicity in ACID compliant is either we get all or nothing of the transaction, which means partial transaction is merely unacceptable.

So if i understand this right, the innodb_timeout handling will corrupt the transaction atomicity!!!
This is FATAL and absolut not acceptable on my option.